actpool/actpool.go

```diff
@@ -155,113 +183,162 @@ func (ap *actPool) AddActionEnvelopeValidators(fs ...action.SealedEnvelopeValida
 // Then starting from the current confirmed nonce, iteratively update pending nonce if nonces are consecutive and pending
 // balance is sufficient, and remove all the subsequent actions once the pending balance becomes insufficient
 func (ap *actPool) Reset() {
-	ap.mutex.Lock()
-	defer ap.mutex.Unlock()
-
 	ap.reset()
 }
 
-func (ap *actPool) ReceiveBlock(*block.Block) error {
-	ap.mutex.Lock()
-	defer ap.mutex.Unlock()
+func (ap *actPool) reset() {
+	var wg sync.WaitGroup
+	for i := range ap.worker {
+		wg.Add(1)
+		go func(worker *queueWorker) {
+			defer wg.Done()
+			worker.Reset()
+		}(ap.worker[i])
+	}
+	wg.Wait()
+}
 
+func (ap *actPool) ReceiveBlock(*block.Block) error {
 	ap.reset()
 	return nil
 }
 
-// PendingActionIterator returns an action interator with all accepted actions
+// PendingActionMap returns an action interator with all accepted actions
 func (ap *actPool) PendingActionMap() map[string][]action.SealedEnvelope {
-	ap.mutex.Lock()
-	defer ap.mutex.Unlock()
-
-	// Remove the actions that are already timeout
-	ap.reset()
-
-	actionMap := make(map[string][]action.SealedEnvelope)
-	for from, queue := range ap.accountActs {
-		actionMap[from] = append(actionMap[from], queue.PendingActs()...)
+	var (
+		wg             sync.WaitGroup
+		actsFromWorker = make([][]*pendingActions, _numWorker)
+	)
+	for i := range ap.worker {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			actsFromWorker[i] = ap.worker[i].PendingAction()
+		}(i)
+	}
+	wg.Wait()
+
+	ret := make(map[string][]action.SealedEnvelope)
+	for _, v := range actsFromWorker {
+		for _, w := range v {
+			ret[w.sender] = w.acts
+		}
 	}
-	return actionMap
+	return ret
 }
 
 func (ap *actPool) Add(ctx context.Context, act action.SealedEnvelope) error {
-	ap.mutex.Lock()
-	defer ap.mutex.Unlock()
-
 	ctx, span := tracer.NewSpan(ctx, "actPool.Add")
 	defer span.End()
 
+	if err := checkSelpData(&act); err != nil {
+		return err
+	}
+
+	if err := ap.checkSelpWithoutState(ctx, &act); err != nil {
+		return err
+	}
+
```

Comment:
structure-wise it is more clear to still call `validate()` here, `selp.Hash()` check can be safely removed from `validate()` (since it is called earlier, for usage in `workingset.go` should also be fine, since `selp.Hash()` is also called somewhere else in there)